### PR TITLE
Refactor sdk task projects

### PR DESCRIPTION
**Pull request by chcosta**
Fixes https://github.com/dotnet/arcade/issues/1452

Refactors SDK task projects so that they are more isolated from repo state.  Previous version was a step in the right direction but didn't isolate enough so that the sdk task projects are only using sdk things (not things from the repo). 

This creates a separate "SdkTask.proj" which is the simplified Build.proj equivalent but used solely to restore / execute the sdk task project.

Also, some minor yaml cleanup.

Validation job is in progress - https://dnceng.visualstudio.com/internal/_build/results?buildId=59023

I'm going to be OOF for a few weeks (though I'll probably iterate a bit on this if the validation job fails just so there's a working option).  Leaving this PR here in case this becomes a blocking issue and someone needs to make progress.

### CODE REVIEW

**Comment by chcosta** [comment]
Validation job is probably going to fail due to publishing contention.  I'll reschedule a build a bit later tonight.

**Comment by chcosta** [comment]
Rerun validation job https://dnceng.visualstudio.com/internal/_build/results?buildId=59207

**Comment by chcosta** [comment]
@tmat , can you take a look?

File: eng/common/build.ps1 [hunk 1]
```diff
@@ -31,14 +29,14 @@ Param(
 function Print-Usage() {
     Write-Host "Common settings:"
     Write-Host "  -configuration <value>  Build configuration: 'Debug' or 'Release' (short: -c)"
-    Write-Host "  -verbosity <value>      Msbuild verbosity: q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic] (short: -v)"
-    Write-Host "  -binaryLog              Output binary log (short: -bl)"
+    Write-Host "  -v[erbosity] <value>    Msbuild verbosity: q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic]"
+    Write-Host "  -[bl|binaryLog]         Output binary log"
```
Comment (tmat) [inline eng/common/build.ps1]: These changes seem irrelevant to the purpose of the PR. 
I'm not sure if I like this way of writing these. The previous version seem more readable.
In any case, forgot `-c[onfiguration]`.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTask.proj [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  Required parameters:
+    RepoRoot                        Repository root.
+    Projects                        List of projects to build. Semicolon separated, may include globs.
+
+  Optional parameters:
+    ContinuousIntegrationBuild      "true" when building on a CI server (PR build or official build)
+
+    Workaround for https://github.com/NuGet/Home/issues/4695
+    QuietRestore                    "true" to suppress Restore output.
+    QuietRestoreBinaryLog           "true" to produce binary log for Restore.
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTask.proj]: Do we need these? I'd just remove everything related to QuietRestore. Will simplify this a lot.

File: eng/common/sdk-task.ps1 [hunk 3]
```diff
@@ -0,0 +1,73 @@
+[CmdletBinding(PositionalBinding=$false)]
+Param(
+  [string] $projects = "",
+  [string][Alias('v')]$verbosity = "minimal",
+  [string] $msbuildEngine = $null,
+  [bool] $warnAsError = $true,
+  [bool] $nodeReuse = $true,
```
Comment (tmat) [inline eng/common/sdk-task.ps1]: Do we need this customization?

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTask.proj [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+          Condition="'$(_RestoreTools)' == 'true' and '$(_QuietRestore)' == 'true'"
+          StandardOutputImportance="normal" />
+
+    <!--
+      Restore built-in tools.
+    -->
+    <MSBuild Projects="Tools.proj"
+             Targets="Restore"
+             Properties="@(_RestoreToolsProps)"
+             Condition="'$(_RestoreTools)' == 'true' and '$(_QuietRestore)' != 'true'"/>
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTask.proj]: Is this needed? I'd say the project (e.g. PublishBuildAssets.proj) should list all dependencies it needs and we should only restore that project, not Tools.proj.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/PublishBuildAssets.proj [hunk 5]
```diff
@@ -8,15 +8,15 @@
   -->
 
   <Import Project="BuildStep.props" />
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/PublishBuildAssets.proj]: ><Import Project="BuildStep.props" /> [](start = 2, length = 36)

Do we need to import this? Seems like the only thing that this brings in is `MicrosoftDotNetMaestroTasksVersion` from DefaultVersions, which we can easily define here. Then we wouldn't need to make stuff in DefaultVersions conditional.

**Comment by tmat** [comment]
I like the direction this is going. I think we can still simplify quite a bit though.

File: eng/common/sdk-task.ps1 (hunk 3 above)
Comment (chcosta) [reply eng/common/sdk-task.ps1]: I think your point is that these sdk-tasks are only going to run in ci or startup cost doesn't really matter for these tasks as much as they do for a build?  That makes sense, I'll trim this out.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/PublishBuildAssets.proj (hunk 5 above)
Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/PublishBuildAssets.proj]: We also need to import `RepoLayout.props`

Comment (tmat) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/PublishBuildAssets.proj]: I think importing RepoLayout is ok.

Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/PublishBuildAssets.proj]: If we move `MicrosoftDotNetMaestroTasksVersion` here, then we need to move much of the rest of "Tools.proj" here so we ensure reference the same version we restore.  I think that would unnecessarily complicate this.

File: eng/common/sdk-task.ps1 (hunk 3 above)
Comment (tmat) [reply eng/common/sdk-task.ps1]: Yes, they will only run in CI. So no-one will pass in these arguments.

---
In reply to: [246116732](https://github.com/dotnet/arcade/pull/1595#discussion_r246116732) [](ancestors = 246116732)

File: src/Microsoft.DotNet.Arcade.Sdk/tools/PublishBuildAssets.proj (hunk 5 above)
Comment (tmat) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/PublishBuildAssets.proj]: I mean, instead of BuildStep.props let's just import RepoLayout.props.

---
In reply to: [246170604](https://github.com/dotnet/arcade/pull/1595#discussion_r246170604) [](ancestors = 246170604)

Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/PublishBuildAssets.proj]: Tools.proj is where the sdk's PackageReference's are defined.  It also defines the RestoreSources.

So, if we move `MicrosoftDotNetMaestroTasksVersion` into SdkTask.proj, then...

1. Packages tools packages will be restored via Tools.proj which reference the DefaultVersion.props file.

2. If you want to update the version which is used in SdkTask.proj, then you have to update the value there and you have to update the value in DefaultVersion.props so that they both restore and use the same version.

**Comment by chcosta** [comment]
@tmat, PTAL

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj [hunk 6]
```diff
@@ -1,27 +1,31 @@
...
-<Project DefaultTargets="PublishBuildAssets">
+<Project Sdk="Microsoft.NET.Sdk" DefaultTargets="PublishBuildAssets">
   <!--
     Required variables:
       ManifestsPath                     Path to file containing manifest files.
       BuildAssetRegistryToken           Token required to insert metadata into BAR.
       MaestroApiEndpoint            	Maestro's endpoint.
   -->
+  <PropertyGroup>
+    <TargetFramework>net462</TargetFramework>
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj]: >net462 [](start = 21, length = 6)

`net472`?

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj [hunk 7]
```diff
@@ -1,27 +1,31 @@
...
+  </PropertyGroup>
+
+  <Import Project="$(MSBuildThisFileDirectory)../RepoLayout.props"/>
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj]: >$(MSBuildThisFileDirectory) [](start = 19, length = 27)

`$(MSBuildThisFileDirectory)` is unnecessary. MSBuild automatically treats import paths relative to the containing file.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.DotNet.Maestro.Tasks" Version="$(MicrosoftDotNetMaestroTasksVersion)" IsImplicitlyDefined="true" />  
+  </ItemGroup>
+
+  <!-- Force sources to DotNetRestoreSources if building offline, discard any sources set by the repo -->
+  <PropertyGroup Condition="'$(DotNetBuildOffline)' == 'true'">
+    <RestoreSources>$(DotNetRestoreSources)</RestoreSources>
+  </PropertyGroup>
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props]: This shouldn't be necessary, since sdk tasks won't be built in Source Build.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+  <PropertyGroup Condition="'$(DotNetBuildOffline)' != 'true'">
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props]: >Condition="'$(DotNetBuildOffline)' != 'true'" [](start = 17, length = 45)

The same here, the condition is not needed.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    <RestoreSources/>
+    <RestoreSources>
+      $(RestoreSources);
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props]:
```
    <RestoreSources/>
    <RestoreSources>
      $(RestoreSources);
```

can be simplified to 

```
    <RestoreSources>
```

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <MicrosoftDotNetMaestroTasksVersion Condition="'$(MicrosoftDotNetMaestroTasksVersion)' == ''">1.0.0-beta.18478.2</MicrosoftDotNetMaestroTasksVersion>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.DotNet.Maestro.Tasks" Version="$(MicrosoftDotNetMaestroTasksVersion)" IsImplicitlyDefined="true" />  
+  </ItemGroup>
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props]: I'd move the PackageReference to the project itself, unless all sdk tasks require this package.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj (hunk 6 above)
Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj]: Sure, grabbed this from https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Tools.proj#L17, I assume there's a reason that is still referencing net462?

File: eng/common/sdk-task.ps1 [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    Write-Host "  -msbuildEngine <value>  Msbuild engine to use to run build ('dotnet', 'vs', or unspecified)."
+    Write-Host ""
+    Write-Host "Command line arguments not listed above are passed thru to msbuild."
+    Write-Host "The above arguments can be shortened as much as to be unambiguous (e.g. -co for configuration, -t for test, etc.)."
+}
+
+function Build {
+  $toolsetBuildProj = InitializeToolset
+
+  $toolsetBuildProj = Join-Path (Split-Path $toolsetBuildProj -Parent) "sdktasks\SdkTask.proj"
```
Comment (tmat) [inline eng/common/sdk-task.ps1]: >sdktasks [](start = 72, length = 8)

I'd use `SdkTasks` or `sdk-tasks` casing. We already have `XUnit`, so I'd prefer the former.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj (hunk 7 above)
Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj]: Hmm... Old habits, Buildtools caused many patterns where this was not true and being explicit addressed a lot of random failures.  Thanks.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props (hunk 10 above)
Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props]: Ack, looks like the code I pulled this from in Arcade has already been updated.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj (hunk 6 above)
Comment (tmat) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj]: I see. That could also be updated to net472.

Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/PublishBuildAssets.proj]: Filed https://github.com/dotnet/arcade/issues/1750

File: src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props (hunk 11 above)
Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/sdktasks/Versions.props]: With all the recent refactoring, nope.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SdkTask.proj [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+    <ItemGroup>
+      <_CommonProps Include="ContinuousIntegrationBuild=$(ContinuousIntegrationBuild)" />
+      <_CommonProps Include="RepoRoot=$(RepoRoot)" />
+    </ItemGroup>
+
+    <ItemGroup Condition="$(_RestoreTools)">
+      <_RestoreToolsProps Include="@(_CommonProps)"/>
+      <_RestoreToolsProps Include="BaseIntermediateOutputPath=$(ArtifactsToolsetDir)Common"/>
+      <_RestoreToolsProps Include="ExcludeRestorePackageImports=true"/>
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SdkTask.proj]: ><_RestoreToolsProps Include="ExcludeRestorePackageImports=true"/> [](start = 6, length = 65)

I don't think we should set this property. This was used for restoring toolset packages, but here we restore packages of the sdk task projects directly.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SdkTask.proj [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    </PropertyGroup>
+
+    <ItemGroup>
+      <_CommonProps Include="ContinuousIntegrationBuild=$(ContinuousIntegrationBuild)" />
+      <_CommonProps Include="RepoRoot=$(RepoRoot)" />
+    </ItemGroup>
+
+    <ItemGroup Condition="$(_RestoreTools)">
+      <_RestoreToolsProps Include="@(_CommonProps)"/>
+      <_RestoreToolsProps Include="BaseIntermediateOutputPath=$(ArtifactsToolsetDir)Common"/>
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SdkTask.proj]: >BaseIntermediateOutputPath [](start = 35, length = 26)

We should set `BaseIntermediateOutputPath` for `Execute` as well.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SdkTask.proj [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+  <Import Project="$(MSBuildThisFileDirectory)Versions.props"/>
+
+  <Target Name="Execute">
+    <Error Text="Property 'Projects' must be specified" Condition="'$(Projects)' == ''"/>
+    <Error Text="File 'global.json' must exist in directory specified by RepoRoot: '$(_RepoRootOriginal)'" Condition="'$(_RepoRootOriginal)' != '' and !Exists('$(RepoRoot)global.json')"/>
+
+    <PropertyGroup>
+      <_RestoreTools>false</_RestoreTools>
+      <_RestoreTools Condition="'$(Restore)' == 'true'">true</_RestoreTools>
+    </PropertyGroup>
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SdkTask.proj]: We can remove `_RestoreTools` and just use `$(Restore)` instead

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SdkTask.proj [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    </ItemGroup>
+
+    <ItemGroup Condition="$(_RestoreTools)">
+      <_RestoreToolsProps Include="@(_CommonProps)"/>
+      <_RestoreToolsProps Include="BaseIntermediateOutputPath=$(ArtifactsToolsetDir)Common"/>
+      <_RestoreToolsProps Include="ExcludeRestorePackageImports=true"/>
+    </ItemGroup>
+
+    <!--
+      Restore built-in tools.
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SdkTask.proj]: >Restore built-in tools. [](start = 6, length = 23)

This should say "Restore projects"

**Comment by tmat** [comment]
A few more comments. LGTM otherwise.

**Comment by chcosta** [comment]
Thanks Tomas, I'll get this merged and then make sure there's another update to modify the Arcade builds to use this

**Comment by chcosta** [comment]
Blocked on https://github.com/dotnet/core-eng/issues/5025

**Comment by chcosta** [comment]
Arcade builds appear to be working, and based on conversations in the team room regarding the status of machines, I'm going to merge this and then keep an eye on builds.
